mitmproxy/net/http/request.py

```diff
@@ -367,13 +367,64 @@ def _get_urlencoded_form(self):
                 pass
         return ()
 
+    def _sanitized_form_data_replacement(self, form_data):
+        # check if unmodified content changes with URL encoding
+        url_decoded_content = urllib.parse.parse_qsl(self.content, keep_blank_values=True)
+        url_reencoded_content = urllib.parse.urlencode(url_decoded_content, doseq=True).encode()
+        if self.content == url_reencoded_content:
+            # encoding the unmodified content doesn't change anything, so we should be good to go on as usual
+            self.content = mitmproxy.net.http.url.encode(form_data).encode()
+        else:
+            # if we are here, we have to alter encoding behavior to mimic the one used in unmodified POST request
+
+            # check if content isn't encoded at all
+            if self.content == urllib.parse.unquote(url_reencoded_content.decode()).encode():
+                # if we land here, the new content mustn't be URL encoded thus we decode the result before setting new content
+                # this is the case for "https://www.facebook.com/ajax/updatestatus.php" but additionally the "=" sign is
+                # omitted on SOME empty keys
+                self.content = urllib.parse.unquote(mitmproxy.net.http.url.encode(form_data)).encode()
+            # check if content isn't encoded at all and equal signs are omitted
+            elif self.content == urllib.parse.unquote(url_reencoded_content.decode()).replace("=&", "&").encode():
```

Comment:
`if self.content == urllib.parse.unquote(url_reencoded_content.decode()).encode()`

If this statement is reached, we already know that self.content isn't encoded "well formed".
Whereby `url_reencoded_content.decode()` has the same result as a call to `mitmproxy.net.http.url.encode(form_data).encode()` if form_data hasn't been modified.

By unquoting `url_reencoded_content.decode()` and comparing it to self.content, we basically check if self.content isn't urlencoded at all. The need to use unquote() has already been explained (no urldecode available, parse_qs not applicable).

So this first special case check could be rewritten to:
`if self.content == urldecode(urlencode(urldecode(self.content))):`
Which resolves to:
`if self.content == urldecode(self.content):`

Now the only thing we have to do in order to handle this case is - instead of setting self.content to:
`self.content = mitmproxy.net.http.url.encode(form_data).encode()`
set it to:
`self.content = urldecode(mitmproxy.net.http.url.encode(form_data)).encode()`
as already described, unquote() is used to mimic urldecode(). So we end up with what is seen in line 385 to handle "special case 1".